controller/share_manager_controller.go

```diff
@@ -777,7 +779,15 @@ func (c *ShareManagerController) createPodManifest(sm *longhorn.ShareManager, an
 // this scenario would lead to the node no longer being able ot be drained
 // since the volume would never be unattached which prevents the engine from being removed
 func (c *ShareManagerController) shouldClaimOwnership(sm *longhorn.ShareManager) bool {
-	shouldClaim := isControllerResponsibleFor(c.controllerID, c.ds, sm.Name, "", sm.Status.OwnerID)
+	// We prefer keeping the owner of the share manager CR to be the same node
+	// where the share manager pod is running on.
+	preferredOwnerID := ""
+	pod, err := c.ds.GetPod(types.GetShareManagerPodNameFromShareManagerName(sm.Name))
+	if err == nil && pod != nil {
+		preferredOwnerID = pod.Spec.NodeName
+	}
+
+	shouldClaim := isControllerResponsibleFor(c.controllerID, c.ds, sm.Name, preferredOwnerID, sm.Status.OwnerID)
 	isCurrentNodeSchedulable := isNodeSchedulable(c.ds, c.controllerID)
 	if shouldClaim {
 		return isCurrentNodeSchedulable
```

Comment:
isCurrentNodeSchedulable is only relevant, if the pod currently does not exists (share-manager-state == error), since you use the pod.spec.NodeName as preferred owner. If the pod already exists the schedulable status is irrelevant.

Reply:
I see. But I think it is still ok that we switch to a schedulable node even if the pod already exists because:
1. Ownership transferring has a very small cost now
2. The next time the sm stops, it will have to switch to the new node anyways
How do you think?

Reply:
Thank you. Fixed